lib/internal/util/debuglog.js

```diff
@@ -115,7 +126,367 @@ function debuglog(set, cb) {
+ */
+
+/**
+ * @typedef {(label: string, traceLabel?: string, args?: any[]) => void} TimerLog
+ */
+
+/**
+ * Debuglog with time fns and support for trace
+ * @param {string} set
+ * @param {(startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog) => void} cb
+ * @returns {{startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog}}
+ */
+function debugWithTimer(set, cb) {
+  set = StringPrototypeToUpperCase(set);
+
+  if (tracesStores === undefined) {
+    tracesStores = { __proto__: null };
+  }
+
+  /**
+   * @type {LogImpl}
+   */
+  function logImpl(label, timeFormatted, args) {
+    const pid = process.pid;
+    const colors = { colors: lazyUtilColors().shouldColorize(process.stderr) };
+    const coloredPID = inspect(pid, colors);
+
+    if (args === undefined)
+      process.stderr.write(format('%s %s %s: %s\n', set, coloredPID, label, timeFormatted));
+    else
+      process.stderr.write(
+        format(
+          '%s %s %s: %s\n',
+          set,
+          coloredPID,
+          label,
+          timeFormatted,
+          ...new SafeArrayIterator(args),
+        ),
+      );
+  }
+
+  const kTraceCategory = buildCategory(StringPrototypeToLowerCase(set));
+  let debugLogCategoryEnabled = false;
+  let traceCategoryEnabled = false;
+  let timerFlags = kNone;
+
+  /**
+   * @type {TimerStart}
+   */
+  function internalStartTimer(logLabel, traceLabel) {
+    time(
+      tracesStores[set],
+      kTraceCategory,
+      'debuglog.time',
+      timerFlags,
+      logLabel,
+      traceLabel,
+    );
+  }
+
+  /**
+   * @type {TimerEnd}
+   */
+  function internalEndTimer(logLabel, traceLabel) {
+    timeEnd(
+      tracesStores[set],
+      kTraceCategory,
+      'debuglog.timeEnd',
+      timerFlags,
+      logImpl,
+      logLabel,
+      traceLabel,
+    );
+  }
+
+  /**
+   * @type {TimerLog}
+   */
+  function internalLogTimer(logLabel, traceLabel, args) {
+    timeLog(
+      tracesStores[set],
+      kTraceCategory,
+      'debuglog.timeLog',
+      timerFlags,
+      logImpl,
+      logLabel,
+      traceLabel,
+      args,
+    );
+  }
+
+  function init() {
+    if (tracesStores[set] === undefined) {
+      tracesStores[set] = new SafeMap();
+    }
+    emitWarningIfNeeded(set);
+    debugLogCategoryEnabled = testEnabled(set);
+    traceCategoryEnabled = isTraceCategoryEnabled(kTraceCategory);
+
+    if (!debugLogCategoryEnabled) {
+      timerFlags |= kSkipLog;
+    }
+
+    if (!traceCategoryEnabled) {
+      timerFlags |= kSkipTrace;
+    }
+
+    // TODO(H4ad): support traceCategory being enabled dynamically
+    if (debugLogCategoryEnabled || traceCategoryEnabled)
+      cb(internalStartTimer, internalEndTimer, internalLogTimer);
+    else
+      cb(noop, noop, noop);
+  }
+
+  /**
+   * @type {TimerStart}
+   */
+  const startTimer = (logLabel, traceLabel) => {
+    init();
+
+    if (debugLogCategoryEnabled || traceCategoryEnabled)
+      return internalStartTimer(logLabel, traceLabel);
```

Comment:
I don't think we need to return the functions invocation results here? It's better to not return them if the return values aren't needed, lest it gives the impression that the return value is important somehow.